README
Pre-installation requirements
=============================

Dependencies:

libfap:
  http://pakettiradio.net/libfap/
  wget http://pakettiradio.net/downloads/libfap/1.5/libfap-1.5.tar.gz
  tar -zxvf libfap-1.5.tar.gz

  cd libfap-1.5
# apply fap patch
  patch -p2 < ./<path_to_tracker_src>/fap_patch.n7nix
  sudo cp src/fap.h /usr/local/include/
  ./configure
  make
  sudo make install
  NOTE: libraries install to /usr/local/lib

libiniparser:
  http://ndevilla.free.fr/iniparser/
  wget http://ndevilla.free.fr/iniparser/iniparser-3.1.tar.gz
  tar -zxvf iniparser-3.1.tar.gz

  cd iniparser
  sudo cp src/iniparser.h  /usr/local/include
  sudo cp src/dictionary.h /usr/local/include
  make
  sudo cp libiniparser.* /usr/local/lib

gpsd: Optional for GPS daemon
  Build controlled by GPSD_ENABLE variable in Makefile
  http://www.catb.org/gpsd/installation.html
  Installed from package.
   - require: gpsd, libgps-dev
  Verify library libgps is installed: ldconfig -p | grep gps
  Test installation with something like this:
  gpsd -D 5 -N -n /dev/ttyUSB0

GTKAPP: Only used by gtk app, see Makefile
  gtk+-2.0
  libgtk2.0-dev

AX25: Only used with AX25 stack
  Build from source http://www.linux-ax25.org/wiki/CVS
  libax25-dev

WEBAPP: Only used by web app, see Makefile
  node.js
    requires openssl libssl-dev
    https://github.com/joyent/node
    git clone git://github.com/ry/node.git

    cd node
    ./configure
    make
    sudo make install
    verify by running: $ node -v && npm -v

  json-c
    https://github.com/json-c/json-c
    git clone git://github.com/json-c/json-c.git

    cd json-c
    sh autogen.sh
    ./configure
    make
    sudo make install
    ldconfig
# if /usr/local/include/json does not exist
 ln -s /usr/local/include/json-c /usr/local/include/json

  javascript libraries
    Needs version 1.8.3 or greater
    wget http://bit.ly/jqsource -O jquery.js
    sudo cp jquery.js /usr/share/dantracker
    Above taken care of by install script.

  node modules
    ctype
    iniparser
    websocket
    connect
    serve-static
    finalhandler

Tools required:
  build-essential (includes libc-dev, make, gcc, ... etc)
  pkg-config
  imagemagick
  git
  automake
  autoconf
  libtool


Basic Installation
------------------

`cd' to the directory containing the source code and type
`make' to compile, link & create png image files.

fap Installation
----------------

Before building the fap library run the fap patch in the tracker
repository against its source tree. You need to supply the directory
path to dantracker source.


cd libfap-1.5
patch -p2 < ./<path_to_tracker_src>/fap_patch.n7nix


For dantracker n7nix ONLY
=========================

Web app install notes
=====================

Once core apps, aprs for tracker & aprs-ax25 for spy are built need to
do the following:

check that this package is installed:
screen

Install node.js & npm (package manager for node)
Install node.js modules globally:
npm -g install <module name>
npm list -g

    ctype
    iniparser
    websocket
    connect

As root run install.sh found in tracker dir.  This will copy binary,
javascript & config files to appropriate directories.

Confirm javascript library jquery-1.8.3.min.js exists here:
/usr/share/dantracker/

Edit config files /etc/tracker/aprs_tracker.ini & /etc/tracker/aprs_spy.ini

change [station] mycall

If you don't have a gps connected change [gps] type from nmea to static
If gps is connected set serial port and set type to nmea

For AX.25 operation change config for [ax25] port = xxx
This should be the same port name used in /etc/ax25/axports

As root run /etc/tracker/tracker-up

As root run screen -ls

should see something similar to:

There are screens on:
	30680.Tracker	(04/15/2013 06:10:32 PM)	(Detached)
	21348.Spy	(04/15/2013 01:00:42 PM)	(Detached)
2 Sockets in /var/run/screen/S-root.

Attach to a screen terminal
---------------------------

screen -r Spy

or

screen -r Tracker


If you are successful in getting these screen sessions running & you
are attached use these 'screen' commands:

list windows:
ctrl a w

change windows
ctrl a 0
ctrl a 1
ctrl a 2

leave session by detaching:
ctrl a d

kill them:
ctrl a \

Open up your browser
Is your browser web socket enabled?
http://caniuse.com/websockets

web url for tracker

<your_machine_name>:8080/tracker.html
or
localhost:8080/tracker.html

web url for spy

<your_machine_name>:8081/spy.html
or
localhost:8081/spy.html

The install script does not setup apps to start on a power up or
reboot.  To have the spy & tracker apps startup at boot time do the
following. As root copy tracker startup script to the init.d dir then
use udpate-rc.d to add the daemon.

su
cd /etc/tracker
cp tracker /etc/init.d/

Add a daemon with sequence of 95

update-rc.d tracker defaults 95

The daemon script supports start, stop, restart & status
For example as root:

/etc/init.d/tracker status


Starting tracker or spy from command line without using script
--------------------------------------------------------------
Make sure tracker or spy is not already running

screen -ls

screen -S Tracker -c /etc/tracker/.screenrc.trk

or

screen -S Spy -c /etc/tracker/.screenrc.spy


Console Programs
================

Stand alone programs displaying to console only.

./aprs-is
---------

./aprs-is -c <path_to_ini_config_file>/aprs.ini

Display packets from an APRS server.
Need to specify in ini file:
[net] server_host_address, server_port and range
[static] lat and long
[gps] type=static


./aprs-ax25
-----------

./aprs-ax25 -c <path_to_ini_config_file>/aprs.ini

Display packets from AX.25 stack
Need to edit compile time define aprs-ax25.c and remove #define
PKT_SOCK_ENABLE.


./fakegps
---------

Continuously display gps GGA & RMC sentences that will be used if
[gps]:type=static is configured.


./faptest < file_containing_aprs_packet
---------------------------------------

Reads a packet from stdin and passes to fap library.



For dantracker kk7ds ONLY
=========================

Configuration
=============
Copy one of the ini files from examples directory to same directory
as aprs executable and rename to aprs.ini


Running dantracker kk7dds edition
=================================
Stress-test by using examples/aprstest.ini (modified as appropriate), which
will attempt to eat the entire APRS-IS world feed without crashing.


Test Programs
-------------

Test stand alone programs displaying to console.

./aprs-is
./aprs-ax25

Test display interface.

Using socket connect with AF_INET
./ui -i with ./uiclient -i [NAME] [VALUE]

Using socket connect wit AF_UNIX
./ui -w with ./uiclient -w [NAME] [VALUE]


Normal operation
----------------

UI socket connect via AF_INET
./ui -i with ./aprs -d 127.0.0.1

UI socket connect via AF_UNIX
./ui -w with ./aprs


ChangeLog
Mon May 25 11:10:19 PDT 2015  Basil Gunn <basil@pacabunga.com>
version: 1.01(181)

    * Update for using libfap 1.5, fixes a couple of seg faults.
      Files modified:
	aprs.c
        aprs.h
        fap_patch.n7nix
	README

    * Define USE_SOCK_PACKET to conditionalize using AF_INET/SOCK_PACKET when
      opening a socket. Kept around in case needed on older versions
      of Linux. Use PF_PACKET/SOCK_RAW instead.
      File modified: aprs-ax25.c

    * Add error check when getting a new json object.
      File modified: apps/trunk/aprs/uiclient.c

Wed Dec 10 12:40:48 PST 2014  Basil Gunn <basil@pacabunga.com>
version: 1.01(180)

    * New version tracking implemented.
      Major & minor version defs are in aprs.h
      Build number evaluated from version control.
      Files modified:
        apps/trunk/aprs/plutracker/aprs.c
        apps/trunk/aprs/plutracker/aprs.h
        apps/trunk/aprs/plutracker/aprs-ax25.c
        apps/trunk/aprs/plutracker/aprs-is.c
        apps/trunk/aprs/plutracker/conf.c
        apps/trunk/aprs/plutracker/Makefile
      File deleted: .build

Mon Dec  8 15:19:19 PST 2014  Basil Gunn <basil@pacabunga.com>

    * Update for using libfap 1.4
      Files modified:
	README
	aprs.c
        fap_patch.n7nix

Mon Dec  8 14:32:51 PST 2014  Basil Gunn <basil@pacabunga.com>

    * Fix websock_port config to use higher port number.
      Files modified:
	examples/aprs_spy.ini
	webapp/spy-frontend.js

Fri Oct 10 10:24:02 PDT 2014  Basil Gunn <basil@pacabunga.com>

    * Make gpsd Makefile conditional for running on a VM.
    * Fix ARM build problem.
      Files modified:
        Makefile
        aprs.c
        aprs.h

    * Fix doc describing dependencies for gpsd.
      File modified: README

    * Fix display of which UI protocol is used.
      Files modified:
        aprs-ax25.c
        uiclient.c

    * Minor cosmetic edits.
      Files modified:
        conf.c
        examples/aprs_tracker.ini
        webapp/spy.html
        webapp/tracker-server.js

Wed Oct  8 14:04:03 PDT 2014  Basil Gunn <basil@pacabunga.com>

    * Add config for period of GPS system time setting.
      Fix gps system time setting for NEMA sentences.
      Add gps device to canned packet loop.
      Files modified:
        README
        aprs.c
        aprs.h
        conf.c
        examples/aprs_tracker.ini

    * Direction NW (e.g. 300) was displayed as N, fix courtesy of dbrooke.
      File modified: util.c

    * Change html server code to support lastest connect() (3.x) version.
      File modified: webapp/tracker-server.js

Sat Oct  4 09:08:28 PDT 2014  Basil Gunn <basil@pacabunga.com>

    * Make project current with UDR dev
      Add support for gpsd

      File added: ipdump.c
      Files modified:
        Makefile
        README
        aprs-ax25.c
        aprs.c
        aprs.h
        ax25dump.c
        conf.c
        examples/aprs_tracker.ini
        monixd.h
        nmea.c
        nmea.h
        uiclient.c
        util.c
        util.h
        webapp/spy-frontend.js
        webapp/spy-server.js
        webapp/tracker-frontend.js
        webapp/tracker-server.js

Sat Jul  6 10:52:27 PDT 2013  Basil Gunn <basil@pacabunga.com>

    * Minimal install script now creates jquery directory if it
      doesn't exist.
    * Comments added.
      File modified: install.sh

    * Changes from Erik PA0ESH for Raspberry Pi install script.
      Many additions & fixes.

ui.c
/* -*- Mode: C; tab-width: 8;  indent-tabs-mode: nil; c-basic-offset: 8; c-brace-offset: -8; c-argdecl-indent: 8 -*- */
/* Copyright 2012 Dan Smith <dsmith@danplanet.com> */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <sys/un.h>
#include <arpa/inet.h>
#include <getopt.h>

#include <gtk/gtk.h>
#include <gdk/gdk.h>
#include <pango/pango.h>

#include "aprs.h"
#include "ui.h"
#include "util.h"

#define BG_COLOR "black"
#define FG_COLOR_TEXT "white"

#define FILL_RED    0x990000FF
#define FILL_GREEN  0x006600FF
#define FILL_ORANGE 0xDD7500FF
#define FILL_BLACK  0x000000FF

#define KEY_LT 65361
#define KEY_UP 65362
#define KEY_RT 65363
#define KEY_DN 65364
#define KEY_ENTER 65293
#define KEY_PGUP 65365
#define KEY_PGDN 65366
#define KEY_ESC  65307

GdkPixbuf *aprs_pri_img;
GdkPixbuf *aprs_sec_img;

#define MAX_WIDTH 720
#define PAD_X 0
#define MARGIN_X 10
#define MARGIN_Y  0

#define WX(x) (MARGIN_X+x)
#define WY(y) (MARGIN_Y+y)

enum {
	TYPE_TEXT_LABEL,
	TYPE_IMAGE,
	TYPE_INDICATOR,
	TYPE_MAX,
};

struct named_element {
	const char *name;
	unsigned int type;
	GtkWidget *widget;
	int (*update_fn)(struct named_element *e, const char *value);
	void *data;
};

struct key_map;
struct layout {
	GtkWidget *window;
	GtkWidget *fixed;

	GtkWidget *sep1;
	GtkWidget *sep2;
	GtkWidget *sep3;
	GtkWidget *sep4;

	GtkWidget *tx;
	GtkWidget *rx;

	unsigned int max;
	unsigned int nxt;
	struct named_element *elements;

	struct key_map *key_map;

	struct {
		int main_selected;
		int last_ui_fd;
	} state;
};

struct element_layout {
	const char *name;
	int x;
	int y;
	int width;
	int height;
};

struct key_map {
	char key;
	int keyval;
	int (*switch_fn)(struct layout *l, struct key_map *km);
};

int switch_to_config(struct layout *l, struct key_map *km);
int switch_to_map(struct layout *l, struct key_map *km);
int switch_to_main(struct layout *l, struct key_map *km);
int main_move_cursor(struct layout *l, struct key_map *km);
int main_beacon(struct layout *l, struct key_map *km);
int main_init_kiss(struct layout *l, struct key_map *km);

struct key_map main_screen[] = {
	{'b',  0,         main_beacon},
	{'c',  0,         switch_to_config},
	{'m',  0,         switch_to_map},
	{'k',  0,         main_init_kiss},
	{0,    KEY_DN,    main_move_cursor},
	{0,    KEY_UP,    main_move_cursor},
	{0,    KEY_PGDN,  main_move_cursor},
	{0,    KEY_PGUP,  main_move_cursor},
	{0x00, 0,      NULL}
};

struct key_map config_screen[] = {
	{0x1B, 0, switch_to_main},
	{0x00, 0, NULL}
};

struct element_layout aprs_info_elements[] = {
	{"AI_ICON",     WX(540), WY(0)},
	{"AI_CALLSIGN", WX(0),  WY(10)},
	{"AI_COURSE",   WX(0),  WY(50), 560},
	{"AI_COMMENT",  WX(0),  WY(80), 560},
	{"AI_DISTANCE", WX(220), WY(15)},
	{NULL, 0, 0}
};

struct element_layout aprs_list_elements[] = {
	{"AL_00", WX(0), 120},
	{"AL_01", WX(0), 150},
	{"AL_02", WX(0), 180},
	{"AL_03", WX(0), 210},

	{"AL_04", WX(320), WY(120)},
	{"AL_05", WX(320), WY(150)},
	{"AL_06", WX(320), WY(180)},
	{"AL_07", WX(320), WY(210)},
	{NULL, 0, 0}
};

struct element_layout gps_info_elements[] = {
	{"G_LATLON",		WX(0), WY(250)},

	{"G_SPD",		WX(0), WY(280)},

	{"G_MYCALL",		WX(0),   WY(310)},
	{"G_LASTBEACON",	WX(170), WY(310)},
	{"G_REASON",		WX(370), WY(310)},
	{"G_SIGBARS",		WX(565), WY(310)},
	{NULL, 0, 0}
};

struct element_layout telemetry_elements[] = {
	{"T_VOLTAGE",	WX(0),   WY(420)},
	{"T_TEMP1",	WX(120), WY(420)},
	{NULL,          0,   0}
};

struct element_layout weather_elements[] = {
	{"WX_NAME",	WX(0),		WY(350), 140},
	{"WX_DIST",	WX(140),	WY(350), 430},
	{"WX_DATA",	WX(0),		WY(380), 560},
	{"WX_COMMENT",	WX(0),		WY(410), 560},
	{"WX_ICON",	WX(540),	WY(345)},
	{NULL,          0,   0},
};

struct element_layout indicator_elements[] = {
	{"I_RX",       0,   0, MAX_WIDTH, 110},
	{"I_DG",       0,   0, MAX_WIDTH, 110},
	{"I_TX",       0, 250, MAX_WIDTH, 340},
	{NULL,         0,   0, 0,   0}
};

struct named_element *get_element(struct layout *l, const char *name)
{
	int i;

	for (i = 0; i < l->nxt; i++)
		if (strcmp(l->elements[i].name, name) == 0)
			return &l->elements[i];
	return NULL;
}

int switch_to_config(struct layout *l, struct key_map *km)
{
	gtk_widget_hide(l->window);
	printf("Switch to config\n");
	return 0;
}

int switch_to_map(struct layout *l, struct key_map *km)
{
	printf("Switch to map\n");
	return 0;
}

int switch_to_main(struct layout *l, struct key_map *km)
{
	gtk_widget_show(l->window);
	printf("Switch to main\n");
	return 0;
}

int main_move_cursor(struct layout *l, struct key_map *km)
{
	char buf[] = "AI_00";
	struct named_element *e;
	GdkColor h_color, n_color;
	char sta[] = "000";

	gdk_color_parse("red", &h_color);
	gdk_color_parse(FG_COLOR_TEXT, &n_color);

	if (l->state.main_selected >= 0) {
		sprintf(buf, "AL_%02i", l->state.main_selected);
		e = get_element(l, buf);
		gtk_widget_modify_fg(e->widget, GTK_STATE_NORMAL, &n_color);
	}

	if (km->keyval == KEY_DN || km->keyval == KEY_PGDN) {
		if (l->state.main_selected < 7)
			l->state.main_selected++;
	} else if (km->keyval == KEY_UP || km->keyval == KEY_PGUP) {
		if (l->state.main_selected > -1)
			l->state.main_selected--;
	}

	snprintf(sta, sizeof(sta), "%d", l->state.main_selected);
	ui_send(l->state.last_ui_fd, UI_MSG_NAME_STATION_INFO, sta);

	if (l->state.main_selected < 0)
		return 0;

	sprintf(buf, "AL_%02i", l->state.main_selected);
	e = get_element(l, buf);
	gtk_widget_modify_fg(e->widget, GTK_STATE_NORMAL, &h_color);

	return 0;
}

int main_beacon(struct layout *l, struct key_map *km)
{
	ui_send(l->state.last_ui_fd, UI_MSG_NAME_BEACON, "");

	return 0;
}

int main_init_kiss(struct layout *l, struct key_map *km)
{
	ui_send(l->state.last_ui_fd, UI_MSG_NAME_KISS, "");

	return 0;
}

struct text_update_ctx {
	char *orig_value;
	const char *curptr;
	int scroll_fn_running;
	struct named_element *e;
};

int scroll_text_label(void *data)
{
	struct text_update_ctx *ctx = data;
	PangoLayout *layout = gtk_label_get_layout(GTK_LABEL(ctx->e->widget));

	if (pango_layout_is_ellipsized(layout))
		ctx->curptr += 20;
	else
		ctx->curptr = ctx->orig_value;

	if ((ctx->curptr - ctx->orig_value) > strlen(ctx->orig_value))
		ctx->curptr = ctx->orig_value;

	gtk_label_set_markup(GTK_LABEL(ctx->e->widget), ctx->curptr);

	return TRUE;
}

void str_rstrip(char *string)
{
	char *ptr = string + strlen(string) - 1;
	while ((*ptr == ' ') && (ptr > string))
		*ptr = '\0';
}

int update_text_label(struct named_element *e, const char *value)
{
	struct text_update_ctx *ctx = e->data;
	PangoLayout *layout;

	if (ctx->orig_value && !strcmp(ctx->orig_value, value))
		return 0;

	free(ctx->orig_value);
	ctx->orig_value = strdup(value);
	ctx->curptr = ctx->orig_value;

	gtk_label_set_markup(GTK_LABEL(e->widget), ctx->orig_value);
	layout = gtk_label_get_layout(GTK_LABEL(e->widget));
	if (pango_layout_is_ellipsized(layout) &&
	    !ctx->scroll_fn_running) {
		printf("Scheduling scroll function for %s\n", e->name);
		ctx->scroll_fn_running = 1;
		g_timeout_add(3000, scroll_text_label, ctx);
	}

	return 0;
}

int make_text_label(struct layout *l,
		    const char *name,
		    const char *initial,
		    const char *font)
{
	struct named_element *e = &l->elements[l->nxt++];
	GdkColor color;
	struct text_update_ctx *ctx;

	gdk_color_parse(FG_COLOR_TEXT, &color);

	if (l->nxt == l->max) {
		printf("Exceeded max elements\n");
		exit(1);
	}

	e->name = name;
	e->type = TYPE_TEXT_LABEL;
	e->widget = gtk_label_new(initial);
	e->update_fn = update_text_label;
	e->data = ctx = calloc(sizeof(*ctx), 1);

	ctx->e = e;

	gtk_misc_set_alignment(GTK_MISC(e->widget), 0, 0);
	gtk_label_set_ellipsize(GTK_LABEL(e->widget), PANGO_ELLIPSIZE_END);

	gtk_widget_show(e->widget);

	gtk_widget_modify_font(e->widget,
			       pango_font_description_from_string(font));
	gtk_widget_modify_fg(e->widget, GTK_STATE_NORMAL, &color);

	return 0;
}

/* The standard APRS icon map looks like this:
 *
 *  | 20 pixels |
 *
 * In other words, 20 pixels of image data and one of a separator line.
 * So, since we scale it up by a factor of APRS_IMG_MULT, we locate an
 * icon by skipping 20*MULT icon pixels and 1*MULT lines.
 */

int update_icon(struct named_element *e, const char *value)
{
	GdkPixbuf *icon;
	int index = value[1] - '!';
	int x = (20 + 1) * APRS_IMG_MULT * (index % 16);
	int y = (20 + 1) * APRS_IMG_MULT * (index / 16);
	GdkPixbuf *source = value[0] == '\\' ? aprs_sec_img : aprs_pri_img;

	if (value[0]) {
		icon = gdk_pixbuf_new_subpixbuf(source,
						x + APRS_IMG_MULT,
						y + APRS_IMG_MULT,
						20 * APRS_IMG_MULT,
						20 * APRS_IMG_MULT);
	} else {
		icon = gdk_pixbuf_new(GDK_COLORSPACE_RGB, 0, 8, 20, 20);
		gdk_pixbuf_fill(icon, FILL_BLACK);
	}
	gtk_image_set_from_pixbuf(GTK_IMAGE(e->widget), icon);
	g_object_unref(icon);

	return 0;
}

int make_icon(struct layout *l, const char *name)
{
	struct named_element *e = &l->elements[l->nxt++];

	e->name = name;
	e->type = TYPE_IMAGE;
	e->widget = gtk_image_new();
	e->update_fn = update_icon;

	gtk_widget_show(e->widget);

	return 0;
}

int update_bars(struct named_element *e, const char *value)
{
	GdkPixbuf *icon;
	int bars = atoi(value);
	char *path = NULL;

	if (asprintf(&path, "images/bars_%i.png", bars) == -1)
		return 0;

	icon = gdk_pixbuf_new_from_file(path, NULL);
	gtk_image_set_from_pixbuf(GTK_IMAGE(e->widget), icon);
	g_object_unref(icon);

	free(path);

	return 0;
}

int make_bars(struct layout *l, const char *name)
{
	struct named_element *e = &l->elements[l->nxt++];

	e->name = name;
	e->type = TYPE_IMAGE;
	e->widget = gtk_image_new();
	e->update_fn = update_bars;

	gtk_widget_show(e->widget);

	return 0;
}

int hide_indicator(void *data)
{
	GtkWidget *widget = data;
	gtk_widget_hide(widget);

	return FALSE; /* Don't run me again */
}

int update_indicator(struct named_element *e, const char *value)
{
	int delay = atoi(value);

	gtk_widget_show(e->widget);
	g_timeout_add(delay, hide_indicator, e->widget);

	return 0;
}

int make_indicator(struct layout *l, const char *name, int color, int height)
{
	struct named_element *e = &l->elements[l->nxt++];
	GdkPixbuf *area;

	e->name = name;
	e->type = TYPE_INDICATOR;
	e->widget = gtk_image_new();
	e->update_fn = update_indicator;

	area = gdk_pixbuf_new(GDK_COLORSPACE_RGB, 0, 8, 720, height);
	gdk_pixbuf_fill(area, color);
	gtk_image_set_from_pixbuf(GTK_IMAGE(e->widget), area);
	g_object_unref(area);

	/* Don't show until asked, unless testing */
	//gtk_widget_show(e->widget);

	return 0;
}

int put_widgets(struct layout *l, struct element_layout *layouts)
{
	int i;

	for (i = 0; layouts[i].name; i++) {
		struct element_layout *el = &layouts[i];
		struct named_element *e = get_element(l, el->name);
		int width = el->width;
		if (width == 0)
			width = MAX_WIDTH - MARGIN_X - el->x;
		gtk_widget_set_size_request(e->widget, width, -1);
		gtk_fixed_put(GTK_FIXED(l->fixed), e->widget,
			el->x + MARGIN_X, el->y + MARGIN_Y);
	}

	return 0;
}

int make_aprs_info(struct layout *l)
{
	make_text_label(l, "AI_CALLSIGN", "N0CALL", "Sans 30");
	make_text_label(l, "AI_COURSE",   "", "Sans 22");
	make_text_label(l, "AI_COMMENT",  "", "Sans 18");
	make_text_label(l, "AI_DISTANCE", "", "Sans 22");
	make_icon(l, "AI_ICON");

	put_widgets(l, aprs_info_elements);

	return 0;
}

int make_aprs_list(struct layout *l)
{
	make_text_label(l, "AL_00",  "", "Monospace 20");
	make_text_label(l, "AL_01",  "", "Monospace 20");
	make_text_label(l, "AL_02",  "", "Monospace 20");
	make_text_label(l, "AL_03",  "", "Monospace 20");

	make_text_label(l, "AL_04",  "", "Monospace 20");
	make_text_label(l, "AL_05",  "", "Monospace 20");
	make_text_label(l, "AL_06",  "", "Monospace 20");
	make_text_label(l, "AL_07",  "", "Monospace 20");

	put_widgets(l, aprs_list_elements);

	return 0;
}

int make_gps_info(struct layout *l)
{
	make_text_label(l, "G_LATLON",  "", "Sans 18");
	make_text_label(l, "G_SPD",     "", "Sans 20");

	make_text_label(l, "G_MYCALL",  "", "Sans 20");
	make_text_label(l, "G_LASTBEACON",  "", "Sans 20");
	make_text_label(l, "G_REASON",  "", "Sans 20");
	make_bars(l, "G_SIGBARS");

	put_widgets(l, gps_info_elements);

	return 0;
}

int make_telemetry(struct layout *l)
{
	make_text_label(l, "T_VOLTAGE",  "", "Sans 18");
	make_text_label(l, "T_TEMP1",  "", "Sans 18");

	put_widgets(l, telemetry_elements);

	return 0;
}

int make_weather(struct layout *l)
{
	make_text_label(l, "WX_NAME", "", "Monospace 20");
	make_text_label(l, "WX_DIST", "", "Sans 18");
	make_text_label(l, "WX_DATA", "", "Sans 18");
	make_text_label(l, "WX_COMMENT", "", "Sans 18");
	make_icon(l, "WX_ICON");

	put_widgets(l, weather_elements);

	return 0;
}

int make_indicators(struct layout *l)
{
	make_indicator(l, "I_TX", FILL_RED, 90);
	make_indicator(l, "I_RX", FILL_GREEN, 110);
	make_indicator(l, "I_DG", FILL_ORANGE, 110);

	put_widgets(l, indicator_elements);

	return 0;
}

gboolean main_button(GtkWidget *window, GdkEvent *event, gpointer *data)
{
	int i;
	struct layout *l = (struct layout *)data;

	for (i = 0; l->key_map[i].switch_fn; i++) {
		struct key_map *km = &l->key_map[i];
		if (km->key && (km->key == *event->key.string)) {
			printf("Dispatching key\n");
			km->switch_fn(l, km);
			return TRUE;
		} else if (km->keyval && (km->keyval == event->key.keyval)) {
			km->switch_fn(l, km);
			return TRUE;
		}
	}

	printf("Key: %s (%i)\n", event->key.string, event->key.keyval);

	return FALSE;
}

int make_window(struct layout *l, int justwindow)
{
	GdkColor color = {0, 0, 0};
	GdkCursor *cursor;
	GdkWindow *window;

	l->window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
	gtk_window_set_default_size(GTK_WINDOW(l->window), MAX_WIDTH, 482);

	l->fixed = gtk_fixed_new();
	gtk_container_add(GTK_CONTAINER(l->window), l->fixed);
	gtk_widget_show(l->fixed);

	gdk_color_parse(BG_COLOR, &color);
	gtk_widget_modify_bg(l->window, GTK_STATE_NORMAL, &color);
	if (!justwindow)
		gtk_window_maximize(GTK_WINDOW(l->window));

	gtk_widget_show(l->window);

	cursor = gdk_cursor_new(GDK_BLANK_CURSOR);
	window = gtk_widget_get_window(GTK_WIDGET(l->window));
	gdk_window_set_cursor(window, cursor);

	make_indicators(l);
	make_aprs_info(l);
	make_aprs_list(l);
	make_gps_info(l);
	make_telemetry(l);
	make_weather(l);

	l->sep1 = gtk_hseparator_new();
	gtk_widget_set_size_request(l->sep1, 720, 10);
	gtk_widget_show(l->sep1);
	gtk_fixed_put(GTK_FIXED(l->fixed), l->sep1, 0, 110);

	l->sep2 = gtk_hseparator_new();
	gtk_widget_set_size_request(l->sep2, 720, 10);
	gtk_widget_show(l->sep2);
	gtk_fixed_put(GTK_FIXED(l->fixed), l->sep2, 0, 240);

	l->sep3 = gtk_hseparator_new();
	gtk_widget_set_size_request(l->sep3, 720, 10);
	gtk_widget_show(l->sep3);
	gtk_fixed_put(GTK_FIXED(l->fixed), l->sep3, 0, 340);

	l->sep4 = gtk_hseparator_new();
	gtk_widget_set_size_request(l->sep4, 720, 10);
	gtk_widget_show(l->sep4);
	gtk_fixed_put(GTK_FIXED(l->fixed), l->sep4, 0, 440);

	gtk_widget_add_events(GTK_WIDGET(l->window), GDK_KEY_PRESS_MASK);
	gtk_signal_connect(GTK_OBJECT(l->window), "key-press-event",
			   G_CALLBACK(main_button), l);

	l->key_map = main_screen;

	return 0;
}

int server_setup_unix(struct state *state)
{
	int sock;
	struct sockaddr_un sockaddr;

	sockaddr.sun_family = AF_UNIX;
	strcpy(sockaddr.sun_path, state->conf.ui_sock_path);
	unlink(state->conf.ui_sock_path);

	sock = socket(AF_UNIX, SOCK_STREAM, 0);
	if (sock < 0) {
		perror("socket");
		return -errno;
	}

	if (bind(sock, (struct sockaddr *)&sockaddr, sizeof(sockaddr))) {
		perror("bind");
		return -errno;
	}

	if (listen(sock, 1)) {
		perror("listen");
		return -errno;
	}

	return sock;
}

int server_setup_inet(struct state *state)
{
	int sock;
	struct sockaddr_in sockaddr;

	sockaddr.sin_family = AF_INET;
	sockaddr.sin_addr.s_addr = 0;
	sockaddr.sin_port = htons(state->conf.ui_inet_port);

	sock = socket(AF_INET, SOCK_STREAM, 0);
	if (sock < 0) {
		perror("socket");
		return -errno;
	}

	if (bind(sock, (struct sockaddr *)&sockaddr, sizeof(sockaddr))) {
		perror("bind");
		return -errno;
	}

	if (listen(sock, 1)) {
		perror("listen");
		return -errno;
	}

	return sock;
}

gboolean server_handle(GIOChannel *source, GIOCondition cond, gpointer user)
{
	int fd;
	struct layout *l = (void *)user;
	struct named_element *e;
	struct ui_msg *msg;
	int ret;

	fd = g_io_channel_unix_get_fd(source);

	ret = ui_get_msg(fd, &msg);
	if (ret < 0) {
		printf("Failed to receive message: %s\n", strerror(-ret));
		return TRUE;
	} else if (ret == 0) {
		printf("Removed client\n");
		close(fd);
		return FALSE;
	}

#if 1
	printf("Setting %s->%s\n", ui_get_msg_name(msg), ui_get_msg_valu(msg));
#endif

	e = get_element(l, ui_get_msg_name(msg));
	if (!e) {
		printf("Unknown element `%s'\n", ui_get_msg_name(msg));
		goto out;
	}

	str_rstrip(ui_get_msg_valu(msg));
	e->update_fn(e, ui_get_msg_valu(msg));
 out:
	free(msg);

	return TRUE;
}

gboolean server_handle_c(GIOChannel *source, GIOCondition cond, gpointer user)
{
	int sock;
	int client;
	struct layout *l = (void *)user;
	struct sockaddr sa;
	unsigned int sa_len;
	GIOChannel *channel;

	sock = g_io_channel_unix_get_fd(source);

	sa_len = sizeof(sa);

	client = accept(sock, &sa, &sa_len);
	if (client < 0) {
		perror("accept");
		return TRUE;
	}

	channel = g_io_channel_unix_new(client);
	g_io_channel_set_encoding(channel, NULL, NULL);
	g_io_add_watch_full(channel, 0, G_IO_IN, server_handle, l, NULL);
	l->state.last_ui_fd = client;

	printf("Added client\n");

	return TRUE;
}

struct opts {
	int window;
	int inet;
};

int server_loop(struct opts *opts, struct layout *l, struct state *state)
{
	int sock;
	GIOChannel *channel;

	if (opts->inet)
		sock = server_setup_inet(state);
	else
		sock = server_setup_unix(state);
	if (sock < 0)
		return sock;

	channel = g_io_channel_unix_new(sock);
	/* read raw binary data without interpretation,
	 * no conversation errors */
	g_io_channel_set_encoding(channel, NULL, NULL);
	g_io_add_watch_full(channel, 0, G_IO_IN, server_handle_c, l, NULL);

	gtk_main();

	return 0;
}

int parse_ui_opts(int argc, char **argv, struct opts *opts)
{
	static struct option lopts[] = {
		{"window",    0, 0, 'w'},
		{"inet",      0, 0, 'i'},
		{NULL,        0, 0,  0 },
	};

	while (1) {
		int c;
		int optidx;

		c = getopt_long(argc, argv, "wi", lopts, &optidx);
		if (c == -1)
			break;

		switch(c) {
		case 'w':
			opts->window = 1;
			break;
		case 'i':
			opts->inet = 1;
		}
	}

	return 0;
}

int main(int argc, char **argv)
{
	struct layout layout;
	struct opts opts;
        struct state state;

        memset(&state, 0, sizeof(state));
	memset(&opts, 0, sizeof(opts));
	memset(&layout, 0, sizeof(layout));
	layout.state.main_selected = -1;

	gtk_init(&argc, &argv);

        parse_ui_opts(argc, argv, &opts);

        if (parse_ini(state.conf.config ? state.conf.config : "aprs.ini", &state)) {
                printf("Invalid config\n");
                exit(1);
        }

	aprs_pri_img = gdk_pixbuf_new_from_file("images/aprs_pri_big.png",
						NULL);
	aprs_sec_img = gdk_pixbuf_new_from_file("images/aprs_sec_big.png",
						NULL);

	layout.max = 32;
	layout.nxt = 0;
	layout.elements = calloc(layout.max, sizeof(struct named_element));

	make_window(&layout, opts.window);

	if (layout.nxt == layout.max) {
		fprintf(stderr, "Too many elements\n");
		abort();
	}

	server_loop(&opts, &layout, &state);
        ini_cleanup(&state);

	return 0;
}
